JASP-Desktop/analysis/analysisform.cpp

```diff
@@ -721,49 +703,59 @@ void AnalysisForm::clearControlError(JASPControlBase* control)
 
 void AnalysisForm::clearFormErrors()
 {
-	if (Utils::currentSeconds() - _lastAddedErrorTimestamp > 5)
-	{
-		_formErrorMessages.clear();
-		_setErrorMessages();
+	_formErrors.clear();
+	emit errorsChanged();
 
-		// Remove also warning without text
-		QSet<JASPControlBase*> controlsWithWarning = _jaspControlsWithWarningSet; // Copy the set: the setHasWarning method changes it.
+	// Remove also warning without text
+	QSet<JASPControlBase*> controlsWithWarning = _jaspControlsWithWarningSet; // Copy the set: the setHasWarning method changes it.
 
-		for (JASPControlBase* control : controlsWithWarning)
-		{
-			if (!_getControlErrorMessageUsingThisJaspControl(control))
-				control->setHasWarning(false);
-		}
+	for (JASPControlBase* control : controlsWithWarning)
+	{
+		if (!_getControlErrorMessageUsingThisJaspControl(control))
+			control->setHasWarning(false);
 	}
 }
 
-void AnalysisForm::formCompletedHandler()
+void AnalysisForm::setAnalysis(QVariant analysis)
 {
-	QTimer::singleShot(0, this, &AnalysisForm::_formCompletedHandler);
+	Analysis * analysisPointer = qobject_cast<Analysis *>(analysis.value<QObject *>());
+
+	if(_analysis == analysisPointer) return;
+
+	if(_analysis && analysisPointer)
+		throw std::runtime_error("An analysis of an analysisform was replaced by another analysis, this is decidedly NOT supported!");
+
+	_analysis = analysisPointer;
+
+	setAnalysisUp();
 }
 
+
+void AnalysisForm::formCompletedHandler()  { QTimer::singleShot(0, this, &AnalysisForm::_formCompletedHandler); }
 void AnalysisForm::_formCompletedHandler()
 {
-	QVariant analysisVariant = QQmlProperty(this, "analysis").read();
-	if (!analysisVariant.isNull())
-	{
-		_analysis	= qobject_cast<Analysis *>(analysisVariant.value<QObject *>());
+	qmlContext(this)->setContextProperty("form", this);
```

Comment:
Is this really necessary?

Reply:
Yes: the form id is known is all forms, but not inside a JASP Widget. Like that it is known everywhere.

Reply:
But I keep seeing warnings in the output `QQmlContext: Cannot set property on internal context.`
I don't know if it is related to this though.

And also, I do not understand "the form id is known is all forms", what it would mean and how it is different for a JASP Widget?

Reply:
I added this in this commit https://github.com/jasp-stats/jasp-desktop/commit/ffb7478a78660c76bf0c141a9c6b5cd60ab13953 to fix overflow issues in the form. I saw that sometimes the GridLayout did not have the form object, and I thought it was because GridLayout is not a JASPControl. But I have just tested it without using the context setting, and it still works. So you can remove it.

Reply:
Ok good!